packages/contracts/contracts/L1/Colosseum.sol

```diff
@@ -319,39 +319,21 @@ contract Colosseum is Initializable, Semver {
         uint256[] calldata _pair
     ) external {
         Types.Challenge storage challenge = challenges[_outputIndex];
-        _validateTurn(challenge);
-
-        bytes32 srcOutputRoot = Hashing.hashOutputRootProof(_srcOutputRootProof);
-        bytes32 dstOutputRoot = Hashing.hashOutputRootProof(_dstOutputRootProof);
 
-        // If asserter timeout, the bisection of segments may not have ended.
-        // Therefore, segment validation only proceeds when bisection is not possible.
-        if (!_isAbleToBisect(challenge)) {
-            require(
-                challenge.segments[_pos] == srcOutputRoot,
-                "Colosseum: the source segment must be matched"
-            );
-            require(
-                challenge.segments[_pos + 1] != dstOutputRoot,
-                "Colosseum: the destination segment must not be matched"
-            );
-        }
+        _validateTurn(challenge);
+        _validateOutputRootProof(_pos, challenge, _srcOutputRootProof, _dstOutputRootProof);
+        _validatePublicInput(_srcOutputRootProof, _dstOutputRootProof, _publicInput, _rlps);
 
-        require(
-            _srcOutputRootProof.nextBlockHash == _dstOutputRootProof.blockHash,
-            "Colosseum: the block hash must be matched"
-        );
+        // TODO(pangssu): waiting for the new Verifier.sol to complete.
+        bytes32 publicInputHash = _hashPublicInput(_srcOutputRootProof.stateRoot, _publicInput);
+        // require(ZK_VERIFIER.verify(_proof, _pair, publicInputHash), "Colosseum: invalid proof");
 
-        // TODO(chokobole): check withdrawal storage root of _dstOutputRootProof against state root of _dstOutputRootProof.
         require(
-            _publicInput.stateRoot == _dstOutputRootProof.stateRoot,
-            "Colosseum: the state root must be matched"
+            !verifiedPublicInputs[publicInputHash],
+            "Colosseum: public input that has already been validated cannot be used again."
         );
```

Comment:
Isn't it better to check if `publicInputHash` is already verified or not before calling `ZK_VERIFIER.verify`?

Reply:
Good. Though it is commented out for now, I moved it down.